### PR TITLE
Add E2E test case for Nutanix cloud provider

### LINKED ISSUE #1138: E2E tests for Nutanix cloud provider
**User Story**
https://github.com/kubermatic/machine-controller/pull/1132 introduces support for Nutanix to machine-controller. We want to implement e2e tests on top of our internal Nutanix setup to test this provider.

**Acceptance Criteria**
An E2E test for Nutanix is added to the prow configuration of machine-controller.

### PR DESCRIPTION
**Pull request by embik**
**What this PR does / why we need it**:

Adding an e2e test case for Nutanix as cloud provider.

To overcome the limitations of the private Nutanix setup, the changes in the ci scripts tunnel the HTTP proxy that is available to the CI infrastructure into the VM via SSH. To bind the forwarded port to a secure IP that is available from a Pod (but not the internet), the Terraform is expanded to include a private network on Hetzner.

The "migrateUID" test scenario is excluded, as the provider implementation is a noop. It does not work well with the HTTP proxy as the given proxy IP is only available inside the VM, not inside the prow job.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #1138

**Special notes for your reviewer**:

**Optional Release Note**:

```release-note
NONE
```

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

### CODE REVIEW

File: .prow.yaml [hunk 1]
```diff
@@ -566,6 +566,29 @@ presubmits:
               memory: 1Gi
               cpu: 500m
 
+  - name: pull-machine-controller-e2e-nutanix
+    optional: true
+    always_run: true
```
Comment (moadqassem) [inline .prow.yaml:571]: I would remove this line here

File: hack/ci-e2e-test.sh [hunk 2]
```diff
@@ -95,6 +98,11 @@ export E2E_SSH_PUBKEY="$(cat ~/.ssh/id_rsa.pub)"
 ./test/tools/integration/provision_master.sh
 
 echo "Running e2e tests..."
+if [[ ! -z "${NUTANIX_E2E_PROXY_HOST:-}" ]]; then
```
Comment (moadqassem) [inline hack/ci-e2e-test.sh:101]: Ok, so this is a bit hacky and might be error prune, however let's go with it for now. The next step would be, introducing connectivity on a higher level. Like running the control plane stack on Nutanix as well.

File: test/e2e/provisioning/all_e2e_test.go [hunk 3]
```diff
@@ -880,6 +881,45 @@ func TestScalewayProvisioningE2E(t *testing.T) {
...
+	project := os.Getenv("NUTANIX_E2E_PROJECT_NAME")
+	subnet := os.Getenv("NUTANIX_E2E_SUBNET_NAME")
+	endpoint := os.Getenv("NUTANIX_E2E_ENDPOINT")
+
+	if password == "" || username == "" || endpoint == "" || cluster == "" || project == "" || subnet == "" {
+		t.Fatal("unable to run the test suite, NUTANIX_E2E_PASSWORD, NUTANIX_E2E_USERNAME, NUTANIX_E2E_CLUSTER_NAME, " +
+			"NUTANIX_E2E_ENDPOINT, NUTANIX_E2E_PROJECT_NAME or NUTANIX_E2E_SUBNET_NAME environment variables cannot be empty")
+	}
+
+	proxyURL := os.Getenv("NUTANIX_E2E_PROXY_URL")
```
Comment (moadqassem) [inline test/e2e/provisioning/all_e2e_test.go:898]: Maybe a bit of documentation about this variable?

File: test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            subnetName: '<< NUTANIX_SUBNET >>'
+            imageName: 'machine-controller-e2e-<< OS_NAME >>'
+            cpus: 2
+            memoryMB: 2048
+            diskSize: 20
+          operatingSystem: "<< OS_NAME >>"
+          operatingSystemSpec:
+            distUpgradeOnBoot: false
+            disableAutoUpdate: true
+            attachSubscription: false
```
Comment (moadqassem) [inline test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml:43]: Why do you need this here? This is for RHEL thus I would remove it

**Review by moadqassem — COMMENTED** [review]
It looks good, just left some minor changes

File: .prow.yaml (hunk 1 above)
Comment (embik) [reply .prow.yaml:571]: All the other private providers like kubevirt and vSphere are set to `always_run`. Shouldn't we keep testing Nutanix as much as possible? Running the test is basically "free" for us.

File: test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml (hunk 4 above)
Comment (embik) [reply test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml:43]: Copied from another `MachineDeployment` testdata YAML. I'll remove it.

File: test/e2e/provisioning/all_e2e_test.go (hunk 3 above)
Comment (embik) [reply test/e2e/provisioning/all_e2e_test.go:898]: Sounds reasonable. Where would you expect the documentation? Inline with the code? Somewhere else?

File: hack/ci-e2e-test.sh (hunk 2 above)
Comment (embik) [reply hack/ci-e2e-test.sh:101]: That's a good suggestion, thanks. The alternative would be to move the machine-controller e2e tests to the same model as the KKP E2E tests and run the control plane in kind, eliminating the VM completely. But I agree this should be seen more as a temporary hack than a permanent solution.

**Pull request merged**